### PR TITLE
[CORE] Add decimal precision tests

### PR DESCRIPTION
**Pull request by ulysses-you**
## What changes were proposed in this pull request?

- Port `DecimalPrecisionSuite` into gluten.
- Add dataType in `ExpressionTransformer` to help check data type

## How was this patch tested?

Pass CI

### CODE REVIEW

File: gluten-core/src/main/scala/org/apache/gluten/backendsapi/SparkPlanExecApi.scala [hunk 1]
```diff
@@ -450,14 +450,6 @@ trait SparkPlanExecApi {
     GenericExpressionTransformer(substraitExprName, children, original)
   }
 
-  def genEqualNullSafeTransformer(
-      substraitExprName: String,
-      left: ExpressionTransformer,
-      right: ExpressionTransformer,
-      original: EqualNullSafe): ExpressionTransformer = {
-    GenericExpressionTransformer(substraitExprName, Seq(left, right), original)
```
Comment (ulysses-you) [inline gluten-core/src/main/scala/org/apache/gluten/backendsapi/SparkPlanExecApi.scala:458]: no backends override this, remove it and use GenericExpressionTransformer

**Comment by ulysses-you** [comment]
cc @jinchengchenghh @baibaichen @PHILO-HE  thank you

File: gluten-core/src/main/scala/org/apache/gluten/expression/ExpressionTransformer.scala [hunk 2]
```diff
@@ -18,6 +18,15 @@ package org.apache.gluten.expression
 
 import org.apache.gluten.substrait.expression.ExpressionNode
 
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.types.DataType
+
 trait ExpressionTransformer {
```
Comment (jinchengchenghh) [inline gluten-core/src/main/scala/org/apache/gluten/expression/ExpressionTransformer.scala:24]: Does this one still should be used? Can we use `ExpressionTransformerWithOrigin` to replace `ExpressionTransformer` and rename `ExpressionTransformerWithOrigin` to `ExpressionTransformer`?

Comment (ulysses-you) [reply gluten-core/src/main/scala/org/apache/gluten/expression/ExpressionTransformer.scala:24]: Some expression transformer did not catch orginal expression. e.g., `BoundReferenceTransformer`, `AttributeReferenceTransformer`, etc. So I leave it in this pr.

Comment (jinchengchenghh) [reply gluten-core/src/main/scala/org/apache/gluten/expression/ExpressionTransformer.scala:24]: We can change it to catch original, do you have other example?
```
val b = bindReference.asInstanceOf[BoundReference]
          AttributeReferenceTransformer(
            a.name,
            b.ordinal,
            a.dataType,
            b.nullable,
            a.exprId,
            a.qualifier,
            a.metadata)

      case b: BoundReference =>
        BoundReferenceTransformer(b.ordinal, b.dataType, b.nullable)
```

Comment (jinchengchenghh) [reply gluten-core/src/main/scala/org/apache/gluten/expression/ExpressionTransformer.scala:24]: For the code clean, I prefer to hold origin for all the `ExpressionTransformer`, what do you think? @zhztheplayer

**Review by yaooqinn — APPROVED** [review]

**Pull request merged**